### DISCUSSION TITLE
Introduce generics to framework

### DISCUSSION THREAD

**Discussion by MariusSpring** [laravel/framework#41597] [Ideas]
Hi!

PhpStorm has recently added support for generics through docblocks. I believe utilizing this will greatly improve developer experience. Take the following example:

**Framework changes:**

```php
if (! function_exists('app')) {
    /**
     * Get the available container instance.
     *
     * @template T
     * 
     * @param class-string<T>|null  $abstract
     * @param array  $parameters
     * @return T|\Illuminate\Contracts\Foundation\Application
     */
    function app($abstract = null, array $parameters = [])
    {
        if (is_null($abstract)) {
            return Container::getInstance();
        }

        return Container::getInstance()->make($abstract, $parameters);
    }
}
```

In app:

```php
// PHP storm (and other IDEs?) is able to autocomplete the `findPopular` function as 
// above we declare that `app` returns an instance of the class string we provide.
$popularPosts = app(BlogPostRepository::class)->findPopular(amount: 10);
```

This could slowely be incorporated bit by bit, however, I believe the most important one is the `app` function for resolving container instances.

Thoughts?

**Comment by MariusSpring** [comment]
Bump.

**Comment by henzeb** [comment]
can't you use the laravel extension instead? https://plugins.jetbrains.com/plugin/7532-laravel

**Reply by pionl** [reply]
It would make PHPStan happy too (at this moment we need to use larastan that "changes" the interfaces for PHPStan).

**Reply by ExpDev07** [reply]
Yes we can, but it would, as pionl said, make static analyzers happy as well.

**Comment by pionl** [comment]
There were some PR but no "desires" to merge them at this moment check https://github.com/laravel/framework/pull/41557 and others.

**Reply by ExpDev07** [reply]
The PR you linked to has been merged, so the example in this thread has been implemented ?

**Reply by pionl** [reply]
It was later on reverted from the code base. There are multiple PR that wanted to fix this before (same week as me).

**Reply by ExpDev07** [reply]
@pionl such a weird thing to revert.

**Comment by shaedrich** [comment]
This particular case is long overdue 👍🏻
